Implemented namespace hash digest

Akashneelesh opened the pull request:
<!-- enter the gh issue after hash -->

- [x] issue #27 
- [ ] follows contribution [guide](https://github.com/keep-starknet-strange/blobstream-starknet/blob/main/CONTRIBUTING.md)
- [x] code change includes tests
- [ ] breaking change

<!-- PR description below -->
Implemented namespace hasher and added tests for the same.

b-j-roberts commented on src/tree/namespace/hasher.cairo:
```diff
@@ -0,0 +1,82 @@
+    append_bytes31(ref bytes, namespace.id);
+    // bytes.concat(@bytes2);
+    bytes.concat(data);
+    let val = bytes.sha256();
```
Computing sha256 hash here and not using resulting value

Akashneelesh replied:
Yep sure

b-j-roberts commented on src/tree/namespace/hasher.cairo:
```diff
@@ -0,0 +1,82 @@
+    } else if (right.min == parity_share_namespace()) {
+        max = left.max;
+    } else {
+        max = namespace_max(left.max, right.max);
```
Doing `namespace_max` calculation twice in this case. Can probably remove the else statement or don't assign `max` before this conditional check,

b-j-roberts commented on src/tree/namespace/merkle_tree.cairo:
```diff
@@ -27,22 +27,16 @@ struct NamespaceMerkleProof {
-//         true
-//     }
-// }
+mod NamespaceMerkleTree {
```
I don't think the above comment should be removed as part of this PR. Also, a lot of the `use` statements added are unneeded, I think.

Akashneelesh replied:
These are actually present as these are required for the merkle tree implementation

b-j-roberts commented on src/tree/namespace/merkle_tree.cairo:
```diff
@@ -11,7 +11,7 @@ struct NamespaceNode {
 struct Namespace {
     version: u8,
-    id: bytes31,
+    id: bytes31, //TODO: #28
```
Comment should be removed since the issue was closed

Akashneelesh replied:
Done

b-j-roberts commented on src/tree/namespace/tests/test_hasher.cairo:
```diff
@@ -0,0 +1,121 @@
+
+    let bytesval: u128 = 0x00000000000000000000000000000000000000000000000000000000;
+
+    let nid: Namespace = Namespace { version: 0x00, id: bytesval.into() };
```
`bytesval` can be declared as bytes31 here, so no need to `into`

b-j-roberts commented on src/tree/namespace/tests/test_hasher.cairo:
```diff
@@ -0,0 +1,121 @@
+        digest: 0x0679246d6c4216de0daa08e5523fb2674db2b6599c3b72ff946b488a15290b62
+    };
+
+    let data = BytesTrait::zero(0);
```
I believe `BytesTrait::new_empty()` should be used for initializing empty bytes.

b-j-roberts commented on src/tree/namespace/tests/test_hasher.cairo:
```diff
@@ -0,0 +1,121 @@
+    let node: NamespaceNode = leaf_digest(nid, @data);
+
+    let res = namespace_node_eq(node, expected);
+    assert(res, 'Not equal to expected digest');
```
The latest pattern for doing asserts in tests is to use the `assert!` macro, which gives better printing outputs with strings.

b-j-roberts commented on src/tree/namespace/tests/test_hasher.cairo:
```diff
@@ -0,0 +1,121 @@
+}
+
+
+impl NamespacePrintTrait of PrintTrait<Namespace> {
```
Can this print trait be removed as well?

b-j-roberts commented on src/tree/tests/test_consts.cairo:
```diff
@@ -7,11 +7,11 @@ fn constants_test() {
     assert!(MAX_HEIGHT == 256, "max height");
     assert!(NODE_PREFIX == 0x01, "node prefix");
 }
+// #[test]
```
Why is this commented out now?

b-j-roberts reviewed (commented):
I left some reviews; please resolve them and also use the `scarb fmt` command for code formatting tests to pass.

Akashneelesh commented on src/tree/namespace/hasher.cairo:
```diff
@@ -0,0 +1,82 @@
+    return r;
+}
+
+impl bytes31PrintImpl of PrintTrait<bytes31> {
```
Yep sure

Akashneelesh commented on src/tree/namespace/hasher.cairo:
```diff
@@ -0,0 +1,82 @@
+    }
+}
+
+fn print_byte_array_as_string(self: @Bytes) {
```
Done

Akashneelesh commented:
> I left some reviews; please resolve them and also use the `scarb fmt` command for code formatting tests to pass.

I have addressed the comments and have resolved it.

b-j-roberts reviewed (approved):
LGTM
